File: src/main/java/io/github/shaksternano/borgar/media/io/reader/ScrimageGifReader.java
package io.github.shaksternano.borgar.media.io.reader;

import com.sksamuel.scrimage.nio.AnimatedGifReader;
import com.sksamuel.scrimage.nio.ImageSource;
import io.github.shaksternano.borgar.media.ImageFrame;
import io.github.shaksternano.borgar.media.MediaUtil;
import io.github.shaksternano.borgar.media.VideoFrame;
import io.github.shaksternano.borgar.media.io.MediaReaderFactory;
import io.github.shaksternano.borgar.util.collection.ClosableIterator;
import io.github.shaksternano.borgar.util.collection.ClosableSpliterator;
import io.github.shaksternano.borgar.util.function.TriFunction;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class ScrimageGifReader extends BaseMediaReader<ImageFrame> {

    private final List<ImageFrame> frames = new ArrayList<>();
    @Nullable
    private Reversed reversed;

    public ScrimageGifReader(File input, String format) throws IOException {
        this(ImageSource.of(input), format);
    }

    public ScrimageGifReader(InputStream input, String format) throws IOException {
        this(ImageSource.of(input), format);
        input.close();
    }

    private ScrimageGifReader(ImageSource imageSource, String format) throws IOException {
        super(format);
        var gif = AnimatedGifReader.read(imageSource);
        frameCount = gif.getFrameCount();
        if (frameCount <= 0) {
            throw new IOException("Could not read any frames!");
        }
        var totalDuration = 0;
        for (var i = 0; i < frameCount; i++) {
            var image = gif.getFrame(i).awt();
            var frameDuration = gif.getDelay(i).toMillis() * 1000;
            frames.add(new ImageFrame(image, frameDuration, totalDuration));
            totalDuration += frameDuration;
        }
        duration = totalDuration;
        frameDuration = (double) duration / frameCount;
        frameRate = 1_000_000 / frameDuration;
        var dimensions = gif.getDimensions();
        width = dimensions.width;
        height = dimensions.height;
    }

    @Override
    public ImageFrame readFrame(long timestamp) {
        return MediaUtil.frameAtTime(timestamp, frames, duration);
    }

    @Override
    public ImageFrame first() {
        return frames.get(0);
    }

    @Override
    public MediaReader<ImageFrame> reversed() {
        if (reversed == null) {
            reversed = new Reversed();
        }
        return reversed;
    }

    @Override
    public ClosableIterator<ImageFrame> iterator() {
        return ClosableIterator.wrap(frames.iterator());
    }

    @Override
    public void forEach(Consumer<? super ImageFrame> action) {
        frames.forEach(action);
    }

    @Override
    public ClosableSpliterator<ImageFrame> spliterator() {
        return ClosableSpliterator.wrap(frames.spliterator());
    }

    @Override
    public void close() {
        frames.clear();
        if (reversed != null) {
            reversed.reversedFrames.clear();
        }
    }

    public enum Factory implements MediaReaderFactory<ImageFrame> {

        INSTANCE;

        @Override
        public MediaReader<ImageFrame> createReader(File media, String format) throws IOException {
            return new ScrimageGifReader(media, format);
        }

        @Override
        public MediaReader<ImageFrame> createReader(InputStream media, String format) throws IOException {
            return new ScrimageGifReader(media, format);
        }
    }

    private class Reversed extends BaseMediaReader<ImageFrame> {

        private final List<ImageFrame> reversedFrames;

        private Reversed() {
            super(ScrimageGifReader.this.format());
            frameCount = ScrimageGifReader.this.frameCount;
            duration = ScrimageGifReader.this.duration;
            frameRate = ScrimageGifReader.this.frameRate;
            frameDuration = ScrimageGifReader.this.frameDuration;
            width = ScrimageGifReader.this.width;
            height = ScrimageGifReader.this.height;
            reversedFrames = reverseFrames(frames, ImageFrame::new);
        }

        private static <E extends VideoFrame<T, E>, T> List<E> reverseFrames(List<E> frames, TriFunction<T, Double, Long, E> frameGenerator) {
            List<E> reversed = new ArrayList<>(frames.size());
            double timestamp = 0;
            for (var i = frames.size() - 1; i >= 0; i--) {
                var frame = frames.get(i);
                reversed.add(frameGenerator.apply(frame.content(), frame.duration(), (long) timestamp));
                timestamp += frame.duration();
            }
            return reversed;
        }

        @Override
        public ImageFrame readFrame(long timestamp) {
            return MediaUtil.frameAtTime(timestamp, reversedFrames, duration);
        }

        @Override
        public MediaReader<ImageFrame> reversed() {
            return ScrimageGifReader.this;
        }

        @Override
        public void close() {
            ScrimageGifReader.this.close();
        }

        @NotNull
        @Override
        public ClosableIterator<ImageFrame> iterator() {
            return ClosableIterator.wrap(reversedFrames.iterator());
        }
    }
}

File: core/src/main/kotlin/io/github/shaksternano/borgar/core/media/reader/ScrimageGifReader.kt
package io.github.shaksternano.borgar.core.media.reader

import com.sksamuel.scrimage.nio.AnimatedGif
import com.sksamuel.scrimage.nio.AnimatedGifReader
import com.sksamuel.scrimage.nio.ImageSource
import io.github.shaksternano.borgar.core.io.DataSource
import io.github.shaksternano.borgar.core.media.ImageFrame
import io.github.shaksternano.borgar.core.media.ImageReaderFactory
import io.github.shaksternano.borgar.core.media.frameAtTime
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.asFlow
import java.io.IOException
import kotlin.time.Duration
import kotlin.time.toKotlinDuration

class ScrimageGifReader(
    gif: AnimatedGif,
) : BaseImageReader() {

    override val frameCount: Int = gif.frameCount
    override val frameRate: Double
    override val duration: Duration
    override val frameDuration: Duration
    override val width: Int
    override val height: Int
    override val loopCount: Int = gif.loopCount
    private val frames: List<ImageFrame>

    init {
        if (frameCount <= 0) throw IOException("Could not read any frames")
        val frames = mutableListOf<ImageFrame>()
        var shortestDuration = Duration.INFINITE
        duration = (0 until frameCount).fold(Duration.ZERO) { total, i ->
            val image = gif.getFrame(i).awt()
            val frameDuration = gif.getDelay(i).toKotlinDuration()
            if (frameDuration < shortestDuration) shortestDuration = frameDuration
            frames.add(ImageFrame(image, frameDuration, total))
            total + frameDuration
        }
        this.frames = frames
        frameDuration = shortestDuration
        frameRate = 1000.0 / frameDuration.inWholeMilliseconds
        val dimensions = gif.dimensions
        width = dimensions.width
        height = dimensions.height
    }

    override suspend fun readFrame(timestamp: Duration): ImageFrame =
        frameAtTime(timestamp, frames, duration)

    override fun asFlow(): Flow<ImageFrame> = frames.asFlow()

    override fun createReversed(): ImageReader = Reversed(this)

    override suspend fun close() = Unit

    private class Reversed(
        private val reader: ScrimageGifReader,
    ) : ReversedImageReader(reader) {

        private val reversedFrames = buildList {
            reader.frames.reversed().fold(Duration.ZERO) { timestamp, frame ->
                add(ImageFrame(frame.content, frame.duration, timestamp))
                timestamp + frame.duration
            }
        }

        override suspend fun readFrame(timestamp: Duration): ImageFrame =
            frameAtTime(timestamp, reversedFrames, duration)

        override fun asFlow(): Flow<ImageFrame> = reversedFrames.asFlow()
    }

    object Factory : ImageReaderFactory {
        override val supportedFormats: Set<String> = setOf("gif")

        override suspend fun create(input: DataSource): ImageReader {
            val bytes = input.toByteArray()
            val imageSource = ImageSource.of(bytes)
            val gif = AnimatedGifReader.read(imageSource)
            return ScrimageGifReader(gif)
        }
    }
}
